### ISSUE TITLE
corrupted double-linked list

### ISSUE THREAD

**Issue by wwj-2017-1117** [ray-project/ray#56770]
### What happened + What you expected to happen

I meet an problem , try to nccl benchmark by ray,   the code as follow:

### Versions / Dependencies

CUDA12.2  pytorch==2.16 ray==2.24

(RayTrainWorker pid=66679) [worker rank=0] local_rank=0 device=cuda:0 world_size=1 CUDA_VISIBLE_DEVICES=0 (RayTrainWorker pid=66679) [rank0]:[W921 01:18:27.451568385 ProcessGroupNCCL.cpp:4561] [PG ID 0 PG GUID 0 Rank 0] using GPU 0 to perform barrier as devices used by this process are currently unknown. This can potentially cause a hang if this rank to GPU mapping is incorrect. Specify device_ids in barrier() to force use of a particular device, or call init_process_group() with a device_id. (RayTrainWorker pid=66679) [worker rank=0] Warmup finished. (RayTrainWorker pid=66679) rank 0 size 1.0MB time 0.0000s bw 120.88GB/s (RayTrainWorker pid=66679) rank 0 size 8.0MB time 0.0001s bw 448.61GB/s (RayTrainWorker pid=66679) corrupted double-linked list (RayTrainWorker pid=66679) *** SIGABRT received at time=1758388709 on cpu 41 *** (RayTrainWorker pid=66679) PC: @ 0x7f994cd139fc (unknown) pthread_kill (RayTrainWorker pid=66679) @ 0x7f994ccbf520 (unknown) (unknown) (RayTrainWorker pid=66679) [2025-09-21 01:18:29,331 E 66679 66679] logging.cc:496: *** SIGABRT received at time=1758388709 on cpu 41 *** (RayTrainWorker pid=66679) [2025-09-21 01:18:29,331 E 66679 66679] logging.cc:496: PC: @ 0x7f994cd139fc (unknown) pthread_kill (RayTrainWorker pid=66679) [2025-09-21 01:18:29,331 E 66679 66679] logging.cc:496: @ 0x7f994ccbf520 (unknown) (unknown) (RayTrainWorker pid=66679) Fatal Python error: Aborted (RayTrainWorker pid=66679) (RayTrainWorker pid=66679) Stack (most recent call first): (RayTrainWorker pid=66679) File "/usr/local/lib/python3.10/site-packages/ray/_private/worker.py", line 953 in main_loop (RayTrainWorker pid=66679) File "/usr/local/lib/python3.10/site-packages/ray/_private/workers/default_worker.py", line 330 in <module> (RayTrainWorker pid=66679)

### Reproduction script

import ray
from ray.train.torch import TorchTrainer
from ray.train import ScalingConfig

import torch
import torch.distributed as dist
import os
import time

def allreduce_benchmark(config):
    rank = int(os.environ["RANK"])
    local_rank = int(os.environ["LOCAL_RANK"])
    world_size = int(os.environ["WORLD_SIZE"])
    torch.cuda.set_device(local_rank)
    device = f"cuda:{local_rank}"
    gpu_id = os.environ.get("CUDA_VISIBLE_DEVICES")
    data_sizes = config.get("data_sizes", [1*1024*1024, 8*1024*1024])
    print(f"[worker rank={rank}] local_rank={local_rank} device={device} world_size={world_size} CUDA_VISIBLE_DEVICES={gpu_id}")

    

    warmup_tensor = torch.ones(2 * 1024 * 1024, device=device)
    dist.barrier()
    dist.all_reduce(warmup_tensor)
    print(f"[worker rank={rank}] Warmup finished.")

    results = []
    for size in data_sizes:
        dist.barrier()
        tensor = torch.ones(size, device=device)
        start = time.perf_counter()
        dist.all_reduce(tensor)
        elapsed = time.perf_counter() - start
        bw = size * 4 / elapsed / 1024 ** 3
        print(f"rank {rank} size {size/1024/1024}MB time {elapsed:.4f}s bw {bw:.2f}GB/s")
        results.append({"rank": rank, "size_mb": size/1024/1024, "elapsed": elapsed, "bandwidth": bw})

    return results

if __name__ == "__main__":
    ray.init(address="auto")

    num_workers_per_node = 8
    num_nodes = 2
    scaling_config = ScalingConfig(
        num_workers=num_workers_per_node * num_nodes,
        use_gpu=True,
        resources_per_worker={"CPU": 1, "GPU": 1},
    )

    config = {
        "data_sizes": [1*1024*1024, 8*1024*1024]
    }

    trainer = TorchTrainer(
        train_loop_per_worker=allreduce_benchmark,
        scaling_config=scaling_config,
        train_loop_config=config,
    )

    result = trainer.fit()
    print("Allreduce benchmark results:", result.metrics)

### Issue Severity

None
